### ISSUE TITLE
Call Julia shared library on Windows (dll) from Python crashes while ok on Linux

### ISSUE THREAD

**Issue by theweaklink** [JuliaLang/PackageCompiler.jl#509]
Hello,

I have been compiling various shared libs on Ubuntu using `create_sysimage()` so that they could be called from Python and so far so good (thanks btw! :+1: ). However once I try to use the same shared libs on Windows, it fails without much clue on the reason why it could be happening. I have reduced the problem to this very bare example.

* function `hello()` to be exposed by the shared lib
```julia
# test_hello.jl
Base.@ccallable function hello()::Cint
    println("hello world");
    return 0;
end
```

* Julia script to build the shared lib from the script above
```julia
# this script simply detects if linux vs win to output the shared lib
# with the proper extension (.so vs .dll)
using Pkg
using PackageCompiler

Pkg.instantiate()

libext = nothing
if Sys.islinux()
    libext = "so"
elseif Sys.iswindows()
    libext = "dll"
end

if libext == nothing
    println("OS type not supported")
else
    create_sysimage(script="test_hello.jl", sysimage_path="libtest_hello.$libext")
end
```

=> shared libs are successfully built on both Ubuntu and WIndows without a hitch

* call `hello()` from Python
```python
# running in REPL within the same dir as libtest_hello.so/.dll
import os
import sys
import ctypes


cur_path = os.getcwd()
if sys.platform == 'linux':
    julia_path = '/opt/julia/julia-1.6.1/'
    julia_lib = os.path.join(julia_path, 'lib', 'julia', 'libjulia-internal.so.1')
    hello_lib = os.path.join(cur_path, 'libtest_hello.so')
    
elif sys.platform == 'win32':
    julia_path = 'C:\\Users\\weaklink\\AppData\\Local\\Programs\\Julia-1.6.1'
    julia_lib = os.path.join(julia_path, 'bin','libjulia-internal.dll')
    hello_lib = os.path.join(cur_path, 'libtest_hello.dll')

if julia_lib is None:
    print('OS not supported')
else:
    # need to load internal julia lib first
    ctypes.CDLL(julia_lib)
    api = ctypes.CDLL(hello_lib, ctypes.RTLD_GLOBAL)
    
    if sys.platform == 'linux':
        api.jl_init_with_image__threading.argtypes = [ctypes.c_char_p, ctypes.c_char_p]
        api.jl_init_with_image__threading(None, bytes(hello_lib, 'utf-8'))
    
    api.hello()
```

On Linux I successfully get:
```bash
<CDLL '/opt/julia/julia-1.6.1/lib/julia/libjulia-internal.so.1', handle c43010 at 0x7fb18ec50b80>
13260992
hello world
0
```

However on Windows I get:
```bash
<CDLL 'C:\Users\weaklink\AppData\Local\Programs\Julia-1.6.1\bin\libjulia-internal.dll', handle 65400000 at 0x16e60509940>

Please submit a bug report with steps to reproduce this fault, and any error messages that follow (in their entirety). Thanks.
Exception: EXCEPTION_ACCESS_VIOLATION at 0x0 -- WaitForSingleObject: (6) The handle is invalid.
```
(note that the Python REPL actually crashes too)

I'm betting that it has something to do with `jl_init_with_image__threading` however I can't find what would be the right call on Windows (I couldn't find it in related GitHub Issues either):
If attempting it I would get the error saying the function doesn't exist:
```bash
>>> api.jl_init_with_image__threading.argtypes = [ctypes.c_char_p, ctypes.c_char_p]
>>> AttributeError: function 'jl_init_with_image__threading' not found
```

(I tried `jl_init()` too)

Does anyone has an idea on what would be going on or how to work around this problem? I'm definitely open to idea and ready to try/test suggestions.
Is that something that would be handled by the `create_library()` PR?

For info:
* I'm running Julia 1.6.1 on both platforms (Ubuntu 20.04 and Windows 10).
* Here is my WIndows Julia `versioninfo()`:
```
Julia Version 1.6.1
Commit 6aaedecc44 (2021-04-23 05:59 UTC)
Platform Info:
  OS: Windows (x86_64-w64-mingw32)
  CPU: Intel(R) Core(TM) i5-8265U CPU @ 1.60GHz
  WORD_SIZE: 64
  LIBM: libopenlibm
  LLVM: libLLVM-11.0.1 (ORCJIT, skylake)
```
* I'm running Python 3.8 on both platforms too

Thanks for any help

**Comment by traversaro** [comment]
I am not sure about your specific problem, but one problem may be that Julia on Windows uses MinGW-w64 (http://mingw-w64.org), while official Python uses Microsoft Visual Studio (MSVC), The two compilers/infrastructures have some points at which they are compatible (such as C ABI) but some other aspects are not compatible (default CRT lib, Standard C++ library). See https://discourse.julialang.org/t/windows-c-binary-compatibility-mingw-vs-msvc/1250/23, https://github.com/JuliaLang/julia/issues/5347 and https://github.com/JuliaInterop/CxxWrap.jl/issues/174 for some related docs.

It would be interesting to try this with a Python interpreter built with the same MinGW-w64 compiler used to build Julia and try if this works, but I am not sure if there is any easy way to do so (probably using the Python binary built by BinaryBuilder?).

**Comment by traversaro** [comment]
> It would be interesting to try this with a Python interpreter built with the same MinGW-w64 compiler used to build Julia and try if the problem is still there, but I am not sure if there is any easy way to do so (probably using the Python binary built by BinaryBuilder?).

Perhaps some BinaryBuilder expert such as @giordano knows something on this?

**Comment by theweaklink** [comment]
Thanks @traversaro I didn't even think of it. Indeed Python on Windows is built with MSVC (Python in Anaconda too).
```python
>>> import sys
>>> sys.version
'3.8.5 (tags/v3.8.5:580fbb0, Jul 20 2020, 15:57:54) [MSC v.1924 64 bit (AMD64)]'
```

Thinking out loud after reading the links you provided, the only viable alternative would be to build Julia and builder packages with MSVC?

**Comment by giordano** [comment]
> the only viable alternative would be to build Julia and builder packages with MSVC?

I wouldn't qualify it a viable option

**Comment by theweaklink** [comment]
yes indeed, let me rephrase to "a hacky solution" :)
I guess the other workaround would be to use PyJulia instead?

That's too bad though, it would be pretty neat to be able to deploy the shared lib cross-platform

**Comment by theweaklink** [comment]
well, well, interestingly enough after some fiddling, calling the `jl_init` on the internal Julia lib does the trick:

```python
    # ... same code as above in original message
    # need to load internal julia lib first
    internal_jl = ctypes.CDLL(julia_lib)
    api = ctypes.CDLL(hello_lib, ctypes.RTLD_GLOBAL)
    
    internal_jl.jl_init_with_image__threading.argtypes = [ctypes.c_char_p, ctypes.c_char_p]
    internal_jl.jl_init_with_image__threading(None, bytes(hello_lib, 'utf-8'))

    api.hello() => "prints "hello world"
```

interesting that on linux it doesn't matter to do either `api.jl_init_with_image__threading()` or `init.jl_init_with_image__threading()`
Let's see if that still work in my complex case depending on artifacts and all. I'll close this issue.

thanks for pointers @traversaro and @giordano
